vscode/microsoft-kiota/src/providers/workspaceTreeProvider.ts

```diff
@@ -1,71 +1,164 @@
+      return [
+        new WorkspaceTreeItem(KIOTA_WORKSPACE_FILE, collapsibleState, 'root')
+      ];
+    }
+
+    if (this.workspaceContent) {
+      if (element.label === KIOTA_WORKSPACE_FILE) {
+        const children: WorkspaceTreeItem[] = [];
+        if (Object.keys(this.workspaceContent.clients).length > 0) {
+          children.push(new WorkspaceTreeItem(CLIENTS, vscode.TreeItemCollapsibleState.Expanded, 'category'));
+        }
+        if (Object.keys(this.workspaceContent.plugins).length > 0) {
+          children.push(new WorkspaceTreeItem(PLUGINS, vscode.TreeItemCollapsibleState.Expanded, 'category'));
+        }
+        if (children.length === 0) {
+          children.push(new WorkspaceTreeItem(vscode.l10n.t("No clients or plugins are available"), vscode.TreeItemCollapsibleState.None, 'info'));
+        }
+        return children;
+      }
+
+      if (isClientType(element.label)) {
+        return Object.keys(this.workspaceContent.clients).map(clientName =>
+          new WorkspaceTreeItem(clientName, vscode.TreeItemCollapsibleState.None, 'item', CLIENTS, this.getProperties(clientName, CLIENTS))
+        );
+      }
+
+      if (isPluginType(element.label)) {
+        return Object.keys(this.workspaceContent.plugins).map(pluginName =>
+          new WorkspaceTreeItem(pluginName, vscode.TreeItemCollapsibleState.None, 'item', PLUGINS, this.getProperties(pluginName, CLIENTS))
+        );
+      }
     }
     return [];
   }
 
-  getTreeItem(element: vscode.TreeItem): vscode.TreeItem {
-    if (element) {
-      element.command = {
-        command: 'kiota.workspace.openWorkspaceFile',
-        title: vscode.l10n.t("Open File"),
-        arguments: [vscode.Uri.file(getWorkspaceJsonPath())]
-      };
-      element.contextValue = 'file';
+  getProperties(name: string, category: string): ClientOrPluginProperties | undefined {
+    if (category && category === CLIENTS) {
+      return this.workspaceContent?.plugins[name];
+    }
+    return this.workspaceContent?.clients[name];
+  }
+
+  getTreeItem(element: WorkspaceTreeItem): WorkspaceTreeItem {
+    if (!element) {
+      return element;
+    }
+
+    switch (element.type) {
+      case 'root':
+        element.command = {
+          command: 'kiota.workspace.openWorkspaceFile',
+          title: vscode.l10n.t("Open File"),
+          arguments: [vscode.Uri.file(getWorkspaceJsonPath())]
+        };
+        element.contextValue = 'folder';
+        break;
+
+      case 'item':
+        const key = element.label;
+        const clientOrPluginKey = this.sharedService.get('clientOrPluginKey');
+        element.iconPath = (clientOrPluginKey && clientOrPluginKey === key) ?
+          new vscode.ThemeIcon('folder-opened') :
+          new vscode.ThemeIcon('folder');
+        break;
     }
     return element;
   }
+
+  public loadWorkspaceContent(): void {
```

Comment:
I believe we're mixing concerns here, and reducing testability in the process.
The tree view providers are responsible for managing the state of the tree view, not for getting the data in the first place. I think we should split things here.

Reply:
This makes sense. I've separated these concerns